### DISCUSSION TITLE
passing array of keys to struct_extract

### DISCUSSION THREAD

**Discussion by dlaplante75** [duckdb/duckdb#14271] [General]
Hello,

Is it possible to pass an array of keys to the struct_extract function so to create a second struct from multiple keys from another struct?Something like snowflake's object_pick function? I've looked around here and in duckdb's documentation but struct_extract seems to only support this one key at a time. 

I'd like to pass an array variable to a prepared statement and create a "key" column by extracting the keys from the array variable from the json column returned by a "select * from '*.json'"

**Comment by soerenwolfers** [comment]
I recently asked for something similar at https://github.com/duckdb/duckdb/discussions/13840

In my understanding, this is not currently possible other than by unnesting the entire struct and then getting the columns you want back via something like
 ```sql
struct_pack(*COLUMNS('regexp'))
```

**Comment by soerenwolfers** [comment]
Here is a solution: 

```sql
CREATE MACRO extract_many(x, y) AS (SELECT struct_pack(*COLUMNS(z -> z in y)) FROM (SELECT unnest(x)));

SELECT 
  y, 
  extract_many(x, ['foo', 'baz'])
FROM
  (SELECT 2 AS y, {'foo': 1, 'bar': 2, 'baz': 3} AS x)
```
```
┌───────┬────────────────────────────────────────────────┐
│   y   │ extract_many(x, main.list_value('foo', 'baz')) │
│ int32 │        struct(foo integer, baz integer)        │
├───────┼────────────────────────────────────────────────┤
│     2 │ {'foo': 1, 'baz': 3}                           │
└───────┴────────────────────────────────────────────────┘
```

**Reply by soerenwolfers** [reply]
This is broken in all sorts of ways though, my quick testing just reveals

**Reply by soerenwolfers** [reply]
https://github.com/duckdb/duckdb/issues/14276

**Comment by pkoppstein** [comment]
You can also go via json, e.g.
   
     select {'a':1,'b':2}::json ->> ['a' ,'b'];
